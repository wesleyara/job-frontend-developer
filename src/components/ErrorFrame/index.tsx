import { MdErrorOutline } from "react-icons/md";

interface ErrorFrameProps {
  error: Error | null;
}

export const ErrorFrame = ({ error }: ErrorFrameProps) => {
  return (
    <section
      data-cy="error-frame"
      className="mt-5 w-full px-3 flex flex-col items-center gap-3"
    >
      <MdErrorOutline size={50} />
      <span className="text-center">{error?.message}</span>
    </section>
  );
};
