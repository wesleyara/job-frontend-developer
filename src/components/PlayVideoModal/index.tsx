import { Dialog, DialogPanel, DialogTitle } from "@headlessui/react";
import { YoutubeDetailsType } from "~/types";
import { Dispatch, SetStateAction } from "react";

interface PlayVideoModalProps {
  item: YoutubeDetailsType;
  isOpen: boolean;
  setIsOpen: Dispatch<SetStateAction<boolean>>;
}

export const PlayVideoModal = ({
  item,
  isOpen,
  setIsOpen,
}: PlayVideoModalProps) => {
  return (
    <Dialog
      open={isOpen}
      onClose={() => setIsOpen(false)}
      className="relative z-50"
    >
      <div className="fixed inset-0 flex mx-auto items-center justify-center p-4">
        <DialogPanel className="space-y-4 border border-gray-300 bg-white p-12 max-w-[768px] w-[90%] text-center rounded-md relative">
          <button
            data-cy="close-youtube-modal"
            className="absolute top-2 right-3"
            onClick={() => setIsOpen(false)}
          >
            X
          </button>
          <DialogTitle className="text-lg">{item.snippet.title}</DialogTitle>

          <iframe
            data-cy="youtube-player"
            src={`https://www.youtube.com/embed/${item.id}`}
            title={item.snippet.title}
            className="inset-0 w-full h-[200px] sm:h-[400px]"
            allowFullScreen
          />
        </DialogPanel>
      </div>
    </Dialog>
  );
};
