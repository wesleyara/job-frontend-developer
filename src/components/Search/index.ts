import { SearchButton } from "./SearchButton";
import { SearchInput } from "./SearchInput";
import { SearchRoot } from "./SearchRoot";

export const Search = {
  Root: SearchRoot,
  Button: SearchButton,
  Input: SearchInput,
};
