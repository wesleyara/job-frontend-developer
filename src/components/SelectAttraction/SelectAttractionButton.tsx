"use client";

import { ButtonHTMLAttributes } from "react";

interface SelectAttractionButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement> {}

export const SelectAttractionButton = ({
  ...rest
}: SelectAttractionButtonProps) => {
  return (
    <button {...rest} data-cy="select-attraction-button" className="btn">
      Selecionar
    </button>
  );
};
