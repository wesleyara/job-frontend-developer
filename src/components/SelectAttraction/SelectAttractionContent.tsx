import { SearchAttraction } from "~/types";
import Image from "next/image";

interface SelectAttractionContentProps {
  item: SearchAttraction;
}

export const SelectAttractionContent = ({
  item,
}: SelectAttractionContentProps) => {
  return (
    <div className="flex rounded-t-md">
      <Image
        src={item.images[0].url}
        alt={item.name}
        width={80}
        height={80}
        className="w-20 h-20 object-cover rounded-t-md"
      />
      <div className="p-4">
        <h3 className="text-md md:text-lg font-semibold">
          {item.name.length > 30
            ? item.name.substring(0, 30) + "..."
            : item.name}
        </h3>
        <p className="text-sm text-gray-500">
          {item.classifications[0].genre.name.length > 20
            ? item.classifications[0].genre.name.substring(0, 20) + "..."
            : item.classifications[0].genre.name}
        </p>
      </div>
    </div>
  );
};
