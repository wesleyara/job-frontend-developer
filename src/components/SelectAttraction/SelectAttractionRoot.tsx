import { ReactNode } from "react";

interface SelectAttractionRootProps {
  children: ReactNode;
}

export const SelectAttractionRoot = ({
  children,
}: SelectAttractionRootProps) => {
  return (
    <div className="flex flex-col border border-gray-300 rounded-md bg-white">
      {children}
    </div>
  );
};
