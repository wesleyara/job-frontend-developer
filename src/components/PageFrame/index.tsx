"use client";

import { useData } from "~/context/DataContext";
import { useState } from "react";
import { FaSearch } from "react-icons/fa";

import { ErrorFrame } from "../ErrorFrame";
import { Search } from "../Search";
import { SelectAttraction } from "../SelectAttraction";
import { ShowAttraction } from "../ShowAttraction";
import { ViewYoutube } from "../ViewYoutube";

export const PageFrame = () => {
  const {
    searchQuery,
    dataTicketSearch,
    dataYoutubeVideosDetails,
    selectedAttraction,
    isLoadingTicketSearch,
    isLoadingYoutubeVideos,
    isLoadingYoutubeVideosDetails,
    errorTicketSearch,
    errorYoutubeVideos,
    errorYoutubeVideosDetails,
    handleChange,
    handleSelectAttraction,
  } = useData();

  const [ratioSelected, setRatioSelected] = useState<string>("");
  const [seeMore, setSeeMore] = useState<boolean>(false);

  const enableToRender = dataTicketSearch && dataTicketSearch.length > 0;

  const render = () => {
    if (isLoadingTicketSearch) {
      return <p className="mt-10">Carregando...</p>;
    }

    if (errorTicketSearch || errorYoutubeVideos || errorYoutubeVideosDetails) {
      return <ErrorFrame error={errorTicketSearch} />;
    }

    if (selectedAttraction) {
      const imagesPerRatio: any = [];

      selectedAttraction.images.forEach(image => {
        const ratio = image.ratio;
        const find = imagesPerRatio.find((item: any) => item.ratio === ratio);

        if (find) {
          find.images.push(image);
        } else {
          imagesPerRatio.push({
            ratio,
            images: [image],
          });
        }
      });

      const selected = imagesPerRatio.find(
        (item: any) => item.ratio === ratioSelected,
      );

      return (
        <>
          <ShowAttraction.Root onClick={() => setSeeMore(prev => !prev)}>
            {!seeMore ? (
              <>
                <ShowAttraction.Info selectedAttraction={selectedAttraction} />
                <ShowAttraction.Badges
                  classifications={selectedAttraction.classifications}
                />
              </>
            ) : (
              <ShowAttraction.Carousel
                setRatioSelected={setRatioSelected}
                selected={selected}
                imagesPerRatio={imagesPerRatio}
                ratioSelected={ratioSelected}
              />
            )}
          </ShowAttraction.Root>

          <span className="flex mt-5">
            {isLoadingYoutubeVideos || isLoadingYoutubeVideosDetails ? (
              <p>Carregando...</p>
            ) : (
              !!dataYoutubeVideosDetails && (
                <ViewYoutube.Root>
                  {dataYoutubeVideosDetails?.map(item => (
                    <ViewYoutube.Card key={item.id} item={item} />
                  ))}
                </ViewYoutube.Root>
              )
            )}
          </span>
        </>
      );
    }

    if (enableToRender) {
      return (
        <div className="flex flex-col gap-3 mt-5 overflow-auto md:w-[600px] w-[90%]">
          {dataTicketSearch.map(item => (
            <SelectAttraction.Root key={item.id}>
              <SelectAttraction.Content item={item} />
              <SelectAttraction.Button
                onClick={() => handleSelectAttraction(item)}
              />
            </SelectAttraction.Root>
          ))}
        </div>
      );
    }
  };

  return (
    <main className="flex flex-col items-center my-24">
      <Search.Root>
        <Search.Input onChange={handleChange} value={searchQuery} />
        <Search.Button icon={FaSearch} />
      </Search.Root>

      {render()}
    </main>
  );
};
