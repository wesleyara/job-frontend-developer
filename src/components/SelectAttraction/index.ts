import { SelectAttractionButton } from "./SelectAttractionButton";
import { SelectAttractionContent } from "./SelectAttractionContent";
import { SelectAttractionRoot } from "./SelectAttractionRoot";

export const SelectAttraction = {
  Root: SelectAttractionRoot,
  Content: SelectAttractionContent,
  Button: SelectAttractionButton,
};
