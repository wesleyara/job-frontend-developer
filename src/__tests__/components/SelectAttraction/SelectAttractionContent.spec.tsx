import { render } from "@testing-library/react";
import { attractionMock } from "~/__tests__/mocks";
import { SelectAttraction } from "~/components/SelectAttraction";
import { beforeEach, describe, expect, it } from "vitest";

describe("SelectAttraction unit test", () => {
  let sut: JSX.Element;

  beforeEach(() => {
    sut = <SelectAttraction.Content item={attractionMock} />;
  });

  it("should be defined", () => {
    expect(sut).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(sut);

    expect(container).toMatchSnapshot();
  });

  it("should have a div element", () => {
    const { container } = render(sut);

    expect(container.querySelector("div")).not.toBeNull();
  });

  it("should have a tailwind class name", () => {
    const { container } = render(sut);

    const element = container.querySelector(".flex");

    expect(element).not.toBeNull();
  });
});
