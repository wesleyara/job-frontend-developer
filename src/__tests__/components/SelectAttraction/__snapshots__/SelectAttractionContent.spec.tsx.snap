// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectAttractionContent unit test > should render correctly 1`] = `
<div>
  <div
    class="flex rounded-t-md items-center"
  >
    <img
      alt="Jota Quest"
      class="w-20 h-20 object-cover rounded-t-md"
      data-nimg="1"
      decoding="async"
      height="80"
      loading="lazy"
      src="/_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=256&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=256&q=75 2x"
      style="color: transparent;"
      width="80"
    />
    <span
      class="px-4"
    >
      <h6
        class="font-semibold"
      >
        Jota Quest
      </h6>
    </span>
  </div>
</div>
`;
